packages/indy-vdr/src/dids/IndyVdrIndyDidRegistrar.ts

```diff
@@ -0,0 +1,329 @@
+    let verkey = options.options.verkey
+    let did = options.did
+    let id
+
+    if (seed && did) {
+      return {
+        didDocumentMetadata: {},
+        didRegistrationMetadata: {},
+        didState: {
+          state: 'failed',
+          reason: `Only one of 'seed' and 'did' must be provided`,
+        },
+      }
+    }
+
+    try {
+      const { namespace, id: submitterId } = parseIndyDid(submitterDid)
+
+      if (did) {
+        id = parseIndyDid(did).id
+        if (!verkey) {
+          return {
+            didDocumentMetadata: {},
+            didRegistrationMetadata: {},
+            didState: {
+              state: 'failed',
+              reason: 'If a did is defined, a matching verkey must be provided',
+            },
+          }
+        }
+        if (!isSelfCertifiedIndyDid(did, verkey)) {
+          throw new Error(`Initial verkey ${verkey} does not match did ˇ${did}`)
+        }
+      } else {
+        // Create a new key and calculate did according to the rules for indy did method
+        const key = await agentContext.wallet.createKey({ seed, keyType: KeyType.Ed25519 })
+        const buffer = Hasher.hash(key.publicKey, 'sha2-256')
+
+        id = TypedArrayEncoder.toBase58(buffer.slice(0, 16))
+        verkey = key.publicKeyBase58
+        did = `did:indy:${namespace}:${id}`
+      }
+
+      // Create base did document
+      const didDocumentBuilder = indyDidDocumentFromDid(did, verkey)
+      let diddocContent
+
+      // Add services if object was passed
+      if (services) {
+        services.forEach((item) => didDocumentBuilder.addService(item))
+
+        const commTypes: CommEndpointType[] = ['endpoint', 'did-communication', 'DIDComm']
+        const serviceTypes = new Set(services.map((item) => item.type))
+
+        const keyAgreementId = `${did}#key-agreement-1`
+
+        // If there is at least a communication service, add the key agreement key
+        if (commTypes.some((type) => serviceTypes.has(type))) {
+          didDocumentBuilder
+            .addVerificationMethod({
+              controller: did,
+              id: keyAgreementId,
+              publicKeyBase58: createKeyAgreementKey(verkey),
+              type: 'X25519KeyAgreementKey2019',
+            })
+            .addKeyAgreement(keyAgreementId)
+        }
+
+        if (!useEndpointAttrib) {
+          // create diddocContent parameter based on the diff between the base and the resulting DID Document
+          diddocContent = deepObjectDiff(
+            didDocumentBuilder.build().toJSON(),
+            indyDidDocumentFromDid(did, verkey).build().toJSON()
+          )
+        }
+      }
+
+      // Build did document
+      const didDocument = didDocumentBuilder.build()
+
+      // If there are services and we are using legacy indy endpoint attrib, make sure they are suitable before registering the DID
+      if (services && useEndpointAttrib) {
+        endpointsAttribFromServices(services)
```

Comment:
It was actually there just to verify services before doing DID registration (because when useEndpointAttrib is set, endpoints are added in a second step), but I finally updated the logic to be more clear.